pkg/api/config/v1alpha1/types.go

```diff
@@ -0,0 +1,84 @@
+/*
+The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	configv1alpha1 "k8s.io/component-base/config/v1alpha1"
+)
+
+// DefaultControllerConfigurationSpec defines the desired state of DefaultControllerConfiguration
+type DefaultControllerConfigurationSpec struct {
+	// SyncPeriod returns the SyncPeriod
+	// +optional
+	SyncPeriod *metav1.Duration `json:"syncPeriod,omitempty"`
+
+	// LeaderElection returns the LeaderElection config
+	// +optional
+	LeaderElection configv1alpha1.LeaderElectionConfiguration `json:"leaderElection,omitempty"`
+
+	// Namespace returns the namespace for the controller
+	// +optional
+	Namespace string `json:"namespace,omitempty"`
+
+	// MetricsBindAddress returns the bind address for the metrics server
+	// +optional
+	MetricsBindAddress string `json:"metricsBindAddress,omitempty"`
+
+	// MetricsBindAddress returns the bind address for the metrics server
+	// +optional
+	Health DefaultControllerConfigurationHealth `json:"health,omitempty"`
+
+	// Port returns the Port for the server
+	// +optional
+	Port *int `json:"port,omitempty"`
+
+	// Host returns the Host for the server
+	// +optional
+	Host string `json:"host,omitempty"`
+
+	// CertDir returns the CertDir
+	// +optional
+	CertDir string `json:"certDir,omitempty"`
+}
+
+// DefaultControllerConfigurationHealth defines the health configs
+type DefaultControllerConfigurationHealth struct {
+	// HealthProbeBindAddress returns the bind address for the health probe
+	// +optional
+	HealthProbeBindAddress string `json:"healthProbeBindAddress,omitempty"`
+
+	// ReadinessEndpointName returns the readiness endpoint name
+	// +optional
+	ReadinessEndpointName string `json:"readinessEndpointName,omitempty"`
+
+	// LivenessEndpointName returns the liveness endpoint name
+	// +optional
+	LivenessEndpointName string `json:"livenessEndpointName,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// DefaultControllerConfiguration is the Schema for the DefaultControllerConfigurations API
+type DefaultControllerConfiguration struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Spec defines the default controller config
+	// +optional
+	Spec DefaultControllerConfigurationSpec `json:"spec,omitempty"`
```

Comment:
Not sure if this is really relevant in this review, but I'm curious :) 

If I wanted to have a custom configuration type, is there a convention in the rest of the Kubernetes community for how to extend the base types?

Would I add a level to the `Spec` hierarchy or would I embed? Or does it matter?

Say I have this:
```go
type MyCustomConfiguration struct {
	metav1.TypeMeta `json:",inline"`
	Spec MyCustomConfigurationSpec `json:"spec,omitempty"
}
```

Would I then have this:
```go
type MyCustomConfigurationSpec struct {
	DefaultControllerConfigurationSpec `json:",inline"`
	MyLoggingConfigurationSpec `json:",inline"`
}
```

Or this:
```go
type MyCustomConfigurationSpec struct {
	Manager DefaultControllerConfigurationSpec `json:"manager,omitempty"`
	Logging MyLoggingConfigurationSpec `json:"logging,omitempty"`
}
```

Reply:
Personally I would do:

```golang
type MyCustomConfigurationSpec struct {
	DefaultControllerConfigurationSpec `json:",inline"`
	MyLoggingConfigurationSpec `json:",inline"`
}
```

But that doesn't mean it's the right approach. I view it as just extending the core, like we do with `ObjectMeta` and `TypeMeta`.

Reply:
Although this leads to an interesting question, I have the funcs for the `ManagerConfig` interface defined on the main Type, currently called `DefaultControllerConfiguration` not on `XXXSpec` so if you embedded this yo'd have to reimplement those funcs.

Do you think that we should cover that case and move the funcs to the `XXXSpec`?

Reply:
@joelanford I actually think I've done this slightly incorrect. At least looking at other componentconfig types like `KubeletConfiguration` the fields should be at the top level so it would be more like this:

```go
type GenericControllerConfiguration struct {
	metav1.TypeMeta `json:",inline"`
	GenericControllerConfigurationSpec `json:",inline"`

        // Custom fields
}
```